scripts/babysitter

```diff
@@ -0,0 +1,364 @@
+
+import argparse
+import subprocess
+import re
+import signal
+import time
+import sys
+import os.path
+import copy
+import tempfile
+import calendar
+import json
+
+### Constants
+
+# Here is how the communication with nunit works. It has to work with two constraints:
+# - We don't invoke nunit directly. We invoke some Makefile which invokes some other Makefile
+#   and at some point down the line someone calls nunit.
+# - nunit has to be able to report back to us even if (especially if) it terminates improperly.
+# To deal with all this, communication babysitter->nunit is done by environment variables,
+# and communication nunit->babysitter is done by leaving behind files in known locations.
+
+# Filenames
+
+CURRENT_TEST_FILE = "babysitter_report_current_test_file.txt"
+RAN_TEST_FILE     = "babysitter_report_ran_test_file.txt"
+FAILED_TEST_FILE  = "babysitter_report_failed_test_file.txt"
+LOGGING_FILE      = "babysitter_report.json_lines"
+
+# Environment keys
+
+CURRENT_TEST_KEY  = 'NUNIT_BABYSITTER_CURRENT_TEST_FILE' # Tell nunit where to leave files
+RAN_TEST_KEY      = 'NUNIT_BABYSITTER_RAN_TEST_FILE'
+FAILED_TEST_KEY   = 'NUNIT_BABYSITTER_FAILED_TEST_FILE'
+RUN_KEY           = 'NUNIT_BABYSITTER_RUN_TEST' # Semicolon-separated list of test names
+RUN_MODE_KEY      = 'NUNIT_BABYSITTER_RUN_MODE' # Equal to either RUN or AFTER
+RETRY_KEY         = 'BABYSITTER_RETRY'          # Equal to an integer
+CRASH_RESUME_KEY  = 'BABYSITTER_CRASH_RESUME'
+VERBOSE_KEY       = 'BABYSITTER_VERBOSE'
+LOGGING_DIR_KEY   = 'WORKSPACE'
+
+# JSON keys
+
+DATE_JSON       = 'date'          # POSIX timestamp of test run
+INVOKE_JSON     = 'invocation'
+COUNT_JSON      = 'passes'        # How many times was command executed?
+LIMIT_JSON      = 'failure_max'
+SUPPORT_JSON    = 'retry_support' # Was the test running with a babysitter-aware nunit?
+FINAL_CODE_JSON = 'final_code'
+TESTS_JSON      = 'tests'         # Holds dictionary of (test name)->(dict with TEST_ keys below)
+TEST_FAILURES         = 'normal_failures'
+TEST_CRASH_FAILURES   = 'crash_failures'
+TEST_TIMEOUT_FAILURES = 'timeout_failures'
+
+### Interpret arguments
+
+scriptname = sys.argv[0]
+
+# This is very silly: This program needs to switch from argparse to manual parsing
+# after the second positional argument, but argparse doesn't let you do this.
+# I work around this with a custom subclass which can selectively swallow errors:
+class Hesitate(Exception):
+	pass
+class HesitantParser(argparse.ArgumentParser):
+	def __init__(s, *args, **kwargs):
+		s.hesitating = True  # Errors will be swallowed until this is set false
+		argparse.ArgumentParser.__init__(s, *args, **kwargs)
+	def error(s, *args, **kwargs):
+		if s.hesitating:
+			raise Hesitate() # Bail out before errors are printed.
+		argparse.ArgumentParser.error(s, *args, **kwargs)
+
+# Define args
+argparser = HesitantParser(description="""\
+Run a test case with a timeout.\n
```

Comment:
Are you really talking about a `test case` (one single test) or rather a `test suite` (a collection of tests) in NUnit terms here?